src/prisma-service.ts

```diff
@@ -0,0 +1,956 @@
+import { PrismaClient } from '@prisma/client';
+import { APIGatewayProxyEvent, APIGatewayProxyResult } from 'aws-lambda';
+import {
+  extractPagination,
+  Pagination,
+  extractTermParam
+} from './request-params';
+import * as TE from 'fp-ts/TaskEither';
+import { pipe } from 'fp-ts/lib/function';
+import { Prisma } from '@prisma/client';
+import {
+  balanceResponse,
+  dagBlockResponse,
+  dagTransactionResponse,
+  dagTransactionsResponse,
+  globalSnapshotResponse,
+  globalSnapshotsResponse,
+  handleError,
+  metagraphBlockResponse,
+  metagraphFeeTransactionResponse,
+  metagraphFeeTransactionsResponse,
+  metagraphSnapshotResponse,
+  metagraphSnapshotsResponse,
+  metagraphsResponse,
+  metagraphTransactionsResponse,
+  metagraphTransactionResponse,
+  missingParameterResponse,
+  notFoundResponse,
+  respond,
+  rewardsResponse
+} from './response';
+import { maxSizeLimit, SearchDirection } from './pagination';
+
+const prisma = new PrismaClient();
+
+const global_snapshot_exists = async (term) => {
+  return await prisma.global_snapshots.findUnique({
+    where: extractTermParam(term),
+    select: { hash: true }
+  });
+};
+
+const metagraph_snapshot_exists = async (metagraph_id, term) => {
+  return await prisma.metagraph_snapshots.findFirst({
+    where: { metagraph_id: metagraph_id, ...extractTermParam(term) },
+    orderBy: {
+      ordinal: 'desc'
+    },
+    select: { hash: true }
+  });
+};
+
+export const paginatedResults = async (
+  event: APIGatewayProxyEvent,
+  nextToCursor,
+  cursorToNext,
+  baseQuery,
+  findMany,
+  transformResponse
+): Promise<APIGatewayProxyResult> => {
+  const task = pipe(
+    extractPagination(event),
+    TE.chain((pagination) =>
+      TE.tryCatch(
+        () =>
+          paginatedQuery(
+            pagination,
+            nextToCursor,
+            cursorToNext,
+            baseQuery,
+            findMany,
+            transformResponse
+          ),
+        (reason) => reason
+      )
+    ),
+    TE.match(
+      (reason) => handleError(reason),
+      (response) => response
+    )
+  );
+  return await task();
+};
+
+const safeNumber = (num, defaultValue) => {
+  return typeof num === 'number' && !isNaN(num) ? num : defaultValue;
+};
+
+const buildPageQuery = <T>(pagination: Pagination<T>, nextToCursor) => {
+  const pageSize = safeNumber(pagination.size, maxSizeLimit);
+  const incrementedSize = pageSize + 1;
+
+  if (
+    pagination &&
+    'searchSince' in pagination &&
+    pagination.searchSince &&
+    'searchDirection' in pagination &&
+    pagination.searchDirection
+  ) {
+    const page =
+      pagination.searchDirection === SearchDirection.Before
+        ? -incrementedSize
```

Comment:
Make sure a negative number is valid here. I think `take` should always be a positive integer.